substratevm/src/com.oracle.objectfile/src/com/oracle/objectfile/pecoff/cv/CVLineRecord.java

```diff
@@ -0,0 +1,205 @@
+        static final int LINE_RECORD_HEADER_SIZE = Integer.BYTES + Short.BYTES * 2 + Integer.BYTES;
+
+        static final int LINE_ENTRY_SIZE = 2 * Integer.BYTES;
+
+        int addr;
+        int lineAndFLags;
+
+        LineEntry(int addr, int line, int deltaEnd, boolean isStatement) {
+            this.addr = addr;
+            assert line <= 0xffffff;
+            assert line >= 0;
+            assert deltaEnd <= 0x7f;
+            assert deltaEnd >= 0;
+            lineAndFLags = line | (deltaEnd << 24) | (isStatement ? 0x80000000 : 0);
+        }
+
+        LineEntry(int addr, int line) {
+            this.addr = addr;
+            this.lineAndFLags = line;
+        }
+
+        int computeContents(byte[] buffer, int initialPos) {
+            int pos = initialPos;
+            pos = CVUtil.putInt(addr, buffer, pos);
+            pos = CVUtil.putInt(lineAndFLags, buffer, pos);
+            return pos;
+        }
+    }
+
+    CVLineRecord(CVDebugInfo cvDebugInfo, String symbolName, PrimaryEntry entry) {
+        super(cvDebugInfo, CVDebugConstants.DEBUG_S_LINES);
+        this.primaryEntry = entry;
+        this.symbolName = symbolName;
+    }
+
+    void addNewFile(FileEntry file) {
+        CVFileRecord fr = cvDebugInfo.getCVSymbolSection().getFileRecord();
+        int fileId = fr.addFile(file);
+        fileBlocks.add(new FileBlock(file, fileId));
+    }
+
+    void addNewLine(int addr, int line) {
+        fileBlocks.get(fileBlocks.size() - 1).addEntry(new LineEntry(addr, line));
+    }
+
+    @Override
+    protected int computeSize(int initialPos) {
+        /* Line record header. */
+        int pos = initialPos + LINE_RECORD_HEADER_SIZE;
+        /* All blocks. */
+        for (FileBlock fileBlock : fileBlocks) {
+            pos = fileBlock.computeSize(pos);
+        }
+        return pos;
+    }
+
+    @Override
+    protected int computeContents(byte[] buffer, int initialPos) {
+        int pos = initialPos;
+
+        assert symbolName != null;
+        /* This implementation can't handle columns yet. */
+        assert !HAS_COLUMNS;
+
+        if (buffer != null) {
+            cvDebugInfo.getCVSymbolSection().markRelocationSite(pos, 4, ObjectFile.RelocationKind.SECREL, symbolName, false, 1L);
+        }
+        pos = CVUtil.putInt(0, buffer, pos);
+
+        if (buffer != null) {
+            cvDebugInfo.getCVSymbolSection().markRelocationSite(pos, 2, ObjectFile.RelocationKind.SECTION, symbolName, false, 1L);
+        }
+        pos = CVUtil.putShort((short) 0, buffer, pos);
+
+        final short flags = HAS_COLUMNS ? CB_HAS_COLUMNS_FLAG : CB_HAS_NO_COLUMNS_FLAG;
+        pos = CVUtil.putShort(flags, buffer, pos); /* flags */
+
+        /*
+         * HighAddr = length of this chunk in object file (fill in correctly later).
+         */
+        final int cbConPos = pos; /* Save position of length(int32). */
+        int highAddr = 0;
+        pos = CVUtil.putInt(highAddr, buffer, pos);
```

Comment:
You could simplify these bits here (and in several other places) by writing the whole header after the content.

For example, if the method `void computeHeader(byte[] buffer, int initialPos, int contentSize)` writes only the header, then writing the content could become something like:
```java
        int pos = initialPos + LINE_RECORD_HEADER_SIZE;
        for (FileBlock fileBlock : fileBlocks) {
            highAddr = Math.max(highAddr, fileBlock.getHighAddr());
            pos = fileBlock.computeContents(buffer, pos);
        }
        computeHeader(initialPos, buffer, highAddr);
        return pos;
```

Reply:
done (here, at least).